### PR TITLE
Add backoff mechanism to profiler agent exporter

### PR DESCRIPTION
**Pull request by Qard**
This adds a backoff mechanism to the profiler agent exporter. I'll put together some tests for it tomorrow. 😸

### CODE REVIEW

File: packages/dd-trace/src/profiling/exporters/agent.js [hunk 1]
```diff
@@ -1,11 +1,58 @@
...
+    const remaining = timeout - (Date.now() - start)
+    if (remaining > 0) await wait(remaining)
+  }
+
+  throw new Error('Profiler agent export back-off failed')
+}
+
+function request (options, body) {
+  return new Promise((resolve, reject) => {
+    const request = http.request(options, resolve)
```
Comment (rochdev) [inline packages/dd-trace/src/profiling/exporters/agent.js:35]: The request function should be cached at the import, otherwise it will be instrumented by the tracer. This is probably already the case which would be a bug.

File: packages/dd-trace/src/profiling/exporters/agent.js [hunk 2]
```diff
@@ -1,11 +1,58 @@
...
+
+async function backoff (base, overallTimeout, task) {
+  for (const timeout of times(base, overallTimeout)) {
+    const start = Date.now()
+    if (await task(timeout)) return
+    const remaining = timeout - (Date.now() - start)
+    if (remaining > 0) await wait(remaining)
+  }
+
+  throw new Error('Profiler agent export back-off failed')
```
Comment (rochdev) [inline packages/dd-trace/src/profiling/exporters/agent.js:30]: I would reword this, as it's not the backoff that failed, it's attempts to send the request that timed out.

File: packages/dd-trace/src/profiling/exporters/agent.js [hunk 3]
```diff
@@ -1,11 +1,58 @@
...
+    const request = http.request(options, resolve)
+    request.on('error', reject)
+    if (body) request.write(body)
+    request.end()
+  })
+}
+
+async function getBody (stream) {
+  const chunks = []
+  for await (const chunk of stream) {
```
Comment (rochdev) [inline packages/dd-trace/src/profiling/exporters/agent.js:44]: I'm getting a bit worried with the increasing number of promises since they tend to cause issues with `async_hooks` and by extension the tracer, and the GC is extremely lazy about them. There was a point in the life of `dd-trace` where there was a single place in the code using promises, and even then it ended up causing performance issues after being called enough times.

While promises are cleaner, in our case we have to prioritize performance and stability over cleaner code.

**Review by rochdev — CHANGES_REQUESTED** [review]

File: packages/dd-trace/src/profiling/exporters/agent.js [hunk 4]
```diff
@@ -1,11 +1,58 @@
...
+
+  // base * n is the current stride length
+  // base * (n - 1) is the sum prior steps
+  // therefore, base * n + base * (n - 1) is the final total
+  while (timeout > base * ((n = 2 ** attempt++) - 1) + base * n) {
+    yield base * n * Math.random()
+  }
+}
+
+async function backoff (base, overallTimeout, task) {
```
[the remark this answers is not in the document]
Comment (Qard) [reply packages/dd-trace/src/profiling/exporters/agent.js:22]: Not in the particular way we want it to work. Specifically this timing randomization algorithm is a requirement to avoid bursty traffic.

File: packages/dd-trace/src/profiling/exporters/agent.js (hunk 2 above)
Comment (Qard) [reply packages/dd-trace/src/profiling/exporters/agent.js:30]: Not necessarily timed out, it's also to recover from other errors. Do you any particular suggestion on how to convey this?

File: packages/dd-trace/src/profiling/exporters/agent.js (hunk 3 above)
Comment (Qard) [reply packages/dd-trace/src/profiling/exporters/agent.js]: I've eliminated more of the promises in the backoff mechanism, though I'm not convinced the performance would actually be noticeably different, especially given that this only runs once every minute. 🤷🏻

File: packages/dd-trace/src/profiling/exporters/agent.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (rochdev) [reply packages/dd-trace/src/profiling/exporters/agent.js]: It looks like `retry` supports randomization with a factor between 1 and 2, and `backoff` supports fully random backoff. This is just from a quick search, but it's likely that there are others as well. Do we have any other requirement not supported by these libraries?

File: packages/dd-trace/src/profiling/exporters/agent.js [hunk 5]
```diff
@@ -51,41 +79,56 @@ class AgentExporter {
...
+      })
 
-        if (res.statusCode >= 400) {
-          return reject(new Error(`Error from the agent: ${res.statusCode}`))
-        }
+      operation.attempt((attempt) => {
+        const timeout = Math.pow(attempt, this._backoffTime)
+        sendRequest({ ...options, timeout }, body, (err, response) => {
+          if (operation.retry(err)) {
+            this._logger.debug(`Error from the agent: ${err.message}`)
```
Comment (rochdev) [inline packages/dd-trace/src/profiling/exporters/agent.js:112]: This should probably be at error level.

File: packages/dd-trace/src/profiling/exporters/agent.js [hunk 6]
```diff
@@ -51,41 +79,56 @@ class AgentExporter {
...
+            return
+          } else if (err) {
+            reject(new Error('Profiler agent export back-off period expired'))
+            return
+          }
+
+          getBody(response, (err, body) => {
+            if (err) {
+              this._logger.debug(`Error reading agent response: ${err.message}`)
```
Comment (rochdev) [inline packages/dd-trace/src/profiling/exporters/agent.js:121]: This should probably be at error level.

File: packages/dd-trace/src/profiling/config.js [hunk 7]
```diff
@@ -45,6 +48,7 @@ class Config {
     )
     this.logger = ensureLogger(options.logger)
     this.flushInterval = flushInterval
+    this.uploadTimeout = uploadTimeout
```
Comment (rochdev) [inline packages/dd-trace/src/profiling/config.js:51]: This doesn't seem to be used anywhere.

File: packages/dd-trace/src/profiling/exporters/agent.js [hunk 8]
```diff
@@ -1,11 +1,39 @@
...
+  })
+}
+
 class AgentExporter {
-  constructor ({ url, logger } = {}) {
+  constructor ({ url, logger, backoffTime, backoffTries } = {}) {
     this._url = url
     this._logger = logger
+    this._backoffTime = backoffTime || 1000
+    this._backoffTries = backoffTries || 6
```
Comment (rochdev) [inline packages/dd-trace/src/profiling/exporters/agent.js:36]: Should this be dynamic? For example, right now with a configured timeout of 1s, there would be a single retry, and for a configured timeout of 300s, there would be no more retries after the 6th one, thus needlessly waiting on an operation that will never succeed.

**Review by rochdev — CHANGES_REQUESTED** [review]

File: packages/dd-trace/src/profiling/exporters/agent.js (hunk 8 above)
Comment (Qard) [reply packages/dd-trace/src/profiling/exporters/agent.js]: Done. It's dynamic now. :)
